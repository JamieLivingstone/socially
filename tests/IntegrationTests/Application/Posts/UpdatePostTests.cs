using System.Linq;
using System.Threading.Tasks;
using Application.Common.Exceptions;
using Application.Posts.Commands.UpdatePost;
using Domain.Entities;
using IntegrationTests.Application.TestUtils;
using NUnit.Framework;

namespace IntegrationTests.Application.Posts
{
  [TestFixture]
  public class UpdatePostTests : TestBase
  {
    [Test]
    public void GivenPostDoesNotExist_ThrowsNotFoundException()
    {
      var command = new UpdatePostCommand
      {
        Slug = "does-not-exist"
      };

      async Task Handler() => await SendAsync(command);

      Assert.ThrowsAsync(typeof(NotFoundException), Handler);
    }

    [Test]
    public void GivenPostAuthorIsNotCurrentUser_ThrowsForbiddenException()
    {
      var target = Seed.Posts().First(c => c.AuthorId != Seed.CurrentUserId);

      var command = new UpdatePostCommand
      {
        Slug = target.Slug
      };

      async Task Handler() => await SendAsync(command);

      Assert.ThrowsAsync(typeof(ForbiddenException), Handler);
    }

    [Test]
    public async Task GivenAValidRequest_UpdatesPost()
    {
      var target = Seed.Posts().First(c => c.AuthorId == Seed.CurrentUserId);

      var command = new UpdatePostCommand
      {
        Slug = target.Slug,
        Title = "Updated title"
      };

      await SendAsync(command);

      var updatedPost = await FindAsync<Post>(p => p.Slug == command.Slug);

      Assert.AreEqual(updatedPost.Title, command.Title);
      Assert.AreEqual(updatedPost.AuthorId, Seed.CurrentUserId);
      Assert.AreEqual(updatedPost.Body, target.Body);
      Assert.AreEqual(target.CreatedAt, updatedPost.CreatedAt);
      Assert.AreNotEqual(target.UpdatedAt, updatedPost.UpdatedAt);
    }
  }
}
