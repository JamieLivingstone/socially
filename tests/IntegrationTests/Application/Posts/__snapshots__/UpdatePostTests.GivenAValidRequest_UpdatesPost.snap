{
  "Id": 1,
  "Slug": "test-post-one",
  "Title": "Updated title",
  "Body": "Test Post One Body",
  "CreatedAt": "2021-01-01T00:00:00",
  "UpdatedAt": "2022-03-18T11:31:46.481778Z",
  "AuthorId": 1,
  "Author": {
    "Id": 1,
    "Name": "Liam Hale",
    "Username": "liam",
    "Email": "liam.hale@test.com",
    "Hash": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==",
    "Salt": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==",
    "Comments": [
      {
        "Id": 1,
        "Message": "Mock comment one",
        "CreatedAt": "0001-01-01T00:00:00",
        "AuthorId": 1,
        "PostId": 1
      },
      {
        "Id": 2,
        "Message": "Mock comment two",
        "CreatedAt": "0001-01-01T00:00:00",
        "AuthorId": 1,
        "PostId": 1
      }
    ],
    "Following": null,
    "Followers": null,
    "Likes": null,
    "Posts": []
  },
  "Comments": [
    {
      "Id": 1,
      "Message": "Mock comment one",
      "CreatedAt": "0001-01-01T00:00:00",
      "AuthorId": 1,
      "Author": {
        "Id": 1,
        "Name": "Liam Hale",
        "Username": "liam",
        "Email": "liam.hale@test.com",
        "Hash": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==",
        "Salt": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==",
        "Comments": [
          {
            "Id": 2,
            "Message": "Mock comment two",
            "CreatedAt": "0001-01-01T00:00:00",
            "AuthorId": 1,
            "PostId": 1
          }
        ],
        "Following": null,
        "Followers": null,
        "Likes": null,
        "Posts": []
      },
      "PostId": 1
    },
    {
      "Id": 2,
      "Message": "Mock comment two",
      "CreatedAt": "0001-01-01T00:00:00",
      "AuthorId": 1,
      "Author": {
        "Id": 1,
        "Name": "Liam Hale",
        "Username": "liam",
        "Email": "liam.hale@test.com",
        "Hash": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==",
        "Salt": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==",
        "Comments": [
          {
            "Id": 1,
            "Message": "Mock comment one",
            "CreatedAt": "0001-01-01T00:00:00",
            "AuthorId": 1,
            "PostId": 1
          }
        ],
        "Following": null,
        "Followers": null,
        "Likes": null,
        "Posts": []
      },
      "PostId": 1
    },
    {
      "Id": 3,
      "Message": "Mock comment three",
      "CreatedAt": "0001-01-01T00:00:00",
      "AuthorId": 2,
      "Author": null,
      "PostId": 1
    },
    {
      "Id": 4,
      "Message": "Mock comment four",
      "CreatedAt": "0001-01-01T00:00:00",
      "AuthorId": 2,
      "Author": null,
      "PostId": 1
    }
  ],
  "Likes": [],
  "PostTags": [
    {
      "PostId": 1,
      "TagId": "programming",
      "Tag": null
    }
  ]
}
